internal/net/grpc/interceptor/server/logging/accesslog.go

```diff
@@ -0,0 +1,152 @@
+//
+// Copyright (C) 2019-2021 vdaas.org vald team <vald@vdaas.org>
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+// Package logging provides gRPC interceptors for access logging
+package logging
+
+import (
+	"context"
+	"path"
+	"time"
+
+	"github.com/vdaas/vald/internal/log"
+	"github.com/vdaas/vald/internal/net/grpc"
+	"github.com/vdaas/vald/internal/observability/trace"
+)
+
+const (
+	grpcKindUnary  = "unary"
+	grpcKindStream = "stream"
+
+	rpcCompletedMessage = "rpc completed"
+)
+
+type AccessLogEntity struct {
+	GRPC      *AccessLogGRPCEntity `json:"grpc,omitempty" yaml:"grpc"`
+	StartTime int64                `json:"startTime,omitempty" yaml:"startTime"`
+	Latency   int64                `json:"latency,omitempty" yaml:"latency"`
+	TraceID   string               `json:"traceID,omitempty" yaml:"traceID"`
+	Error     error                `json:"error,omitempty" yaml:"error"`
+}
+
+type AccessLogGRPCEntity struct {
+	Kind    string `json:"kind,omitempty" yaml:"kind"`
+	Service string `json:"service,omitempty" yaml:"service"`
+	Method  string `json:"method,omitempty" yaml:"method"`
+}
+
+func AccessLogInterceptor() grpc.UnaryServerInterceptor {
+	return func(
+		ctx context.Context,
+		req interface{},
+		info *grpc.UnaryServerInfo,
+		handler grpc.UnaryHandler,
+	) (resp interface{}, err error) {
+		var traceID string
+
+		span := trace.FromContext(ctx)
+		if span != nil {
+			traceID = span.SpanContext().TraceID.String()
+		}
+
+		start := time.Now()
+
+		resp, err = handler(ctx, req)
+
+		latency := time.Since(start).Nanoseconds()
+		startTime := start.UnixNano()
+
+		service, method := parseMethod(info.FullMethod)
+
+		entity := &AccessLogEntity{
+			GRPC: &AccessLogGRPCEntity{
+				Kind:    grpcKindUnary,
+				Service: service,
+				Method:  method,
+			},
+			StartTime: startTime,
```

Comment:
I think there are cases where you want to record the end time like the CreateIndex phase.
You can calculate end time from startTime and latency, but since time.Now is called internally in time.Since, you should record the end time with time.Now and then calculate Latency using time.Sub. How about changing it?

Reply:
I've just added `endTime` field. please check it.

Reply:
checked ! thanks!.